File: src/Systems/Swift/example1.swift
import Foundation

// Abstract Product
protocol Button {
    func render()
}

protocol Checkbox {
    func render()
}

// Concrete Product
class DarkButton: Button {
    func render() {
        print("Dark Button")
    }
}

class LightButton: Button {
    func render() {
        print("Light Button")
    }
}

class DarkCheckbox: Checkbox {
    func render() {
        print("Dark Checkbox")
    }
}

class LightCheckbox: Checkbox {
    func render() {
        print("Light Checkbox")
    }
}

// Abstract Factory
protocol UIFactory {
    func createButton() -> Button
    func createCheckbox() -> Checkbox
}

// Concrete Factory
class DarkFactory: UIFactory {
    func createButton() -> Button {
        return DarkButton()
    }
    func createCheckbox() -> Checkbox {
        return DarkCheckbox()
    }
}

class LightFactory: UIFactory {
    func createButton() -> Button {
        return LightButton()
    }
    func createCheckbox() -> Checkbox {
        return LightCheckbox()
    }
}

func createUIComponents(factory: UIFactory) {
    let button = factory.createButton()
    let checkbox = factory.createCheckbox()
    button.render()
    checkbox.render()
}

// Usage
createUIComponents(factory: DarkFactory())
createUIComponents(factory: LightFactory())

File: src/Systems/Go/example1.go
package main

import "fmt"

// Abstract Product
type Button interface {
    Render()
}

type Checkbox interface {
    Render()
}

// Concrete Product
type DarkButton struct{}

func (d DarkButton) Render() {
    fmt.Println("Dark Button")
}

type LightButton struct{}

func (l LightButton) Render() {
    fmt.Println("Light Button")
}

type DarkCheckbox struct{}

func (d DarkCheckbox) Render() {
    fmt.Println("Dark Checkbox")
}

type LightCheckbox struct{}

func (l LightCheckbox) Render() {
    fmt.Println("Light Checkbox")
}

// Abstract Factory
type UIFactory interface {
    CreateButton() Button
    CreateCheckbox() Checkbox
}

// Concrete Factory
type DarkFactory struct{}

func (d DarkFactory) CreateButton() Button {
    return DarkButton{}
}

func (d DarkFactory) CreateCheckbox() Checkbox {
    return DarkCheckbox{}
}

type LightFactory struct{}

func (l LightFactory) CreateButton() Button {
    return LightButton{}
}

func (l LightFactory) CreateCheckbox() Checkbox {
    return LightCheckbox{}
}

func createUIComponents(factory UIFactory) {
    button := factory.CreateButton()
    checkbox := factory.CreateCheckbox()
    button.Render()
    checkbox.Render()
}

func main() {
    darkFactory := DarkFactory{}
    lightFactory := LightFactory{}
    createUIComponents(darkFactory)
    createUIComponents(lightFactory)
}
